proposals/0540-vision-for-layout-conformance-parity.md

```diff
@@ -0,0 +1,238 @@
+
+### Yoga 2 (flexlayout)
+
+Yoga 2/flexlayout (not to be confused with a popular set of Yoga bindings also named FlexLayout) is a currently closed source spiritual successor to Yoga, developed by Yoga’s owning team. It is undergoing active rollout within Meta, being used internally for [ComponentKit](https://componentkit.org/), and soon [Litho](https://fblitho.com/). It is derived from Yoga, but offers a simplified Modern C++ codebase, and a programming model which lets the consuming application define its own node structure (Data can be embedded into the ShadowNode instead of a separate Yoga Node tree). Aligning to it would allow unifying resources between teams, which is usually a positive.
+
+Despite these advantages, we’ve chosen not to prioritize migration to Yoga 2. Its future roadmap and maintenance is uncertain as is its role in OSS. It has not shown performance benefits over Yoga, and does not currently implement new features. It is still being stabilized, and is not yet at full parity which Yoga (e.g. as of writing, the `flex` shorthand is unsupported). Some of the goals behind Yoga 2 conflict with the goals of React Native. Yoga 2 was built for environments with a high sensitivity to binary size, where we are more interested in adding new complexity, negatively impacting that goal. Yoga 2 does not support code which assumes the bugs and legacy flags of Yoga 1, and we would need to implement changes in the engine to support the current RN ecosystem. Migration in RN would itself be a very large effort blocking enablement of the innovations we would like. We will continue to reevaluate usage of Yoga 2 as it matures, and our ecosystem becomes more compliant, but should not block improving React Native and our ecosystem in the meantime.
+
+### Browser engines
+
+It is tempting to try to reuse an already developed styling engine from a browser. Major advantages are conformance and capabilities, but these are tied to internals of each browser, and add a large footprint and complexity that is undesirable. We would also need to build in Yoga bug compatibility. It would effectively entail rebuilding a new layout engine, heavily forked from a one-time snapshot of a browser engine.
+
+### Taffy
+
+[Taffy](https://github.com/DioxusLabs/taffy) is a Flexbox layout engine written in Rust, authored by Emil Sjölander, a main contributor of Yoga. It is seeing active development under new hands, and has some capabilities that Yoga doesn’t (e.g. it recently added `align-items: 'space-evenly'`). It does not have bug compatibility with the existing RN ecosystem, its rollout/production usage is unknown, and being owned outside of Meta limits our ability to direct the engine in ways which are targeted towards RN’s use-cases.
```

Comment:
Wow this is super helpful to know. Taffy targeting RN quirks (and feedback for hitting RN quirks) seems like a viable way to reuse a lot of the great work you all have done, and to share the work being done by folks contributing to RN. 

Beyond specific project ownership, a benefit of RN having primary ownership of a layout engine is that we can make choices in accordance to the priorities of RN. But that might be indexing too much into the specific case where RN was less aligned to the goals of Yoga 2/FlexLayout. I would definitely be interested in learning more about how RN’s usage aligns with Taffy’s goals.

Apart from the ideological goal, there’s a practical question of “how do we build enough confidence the behavior is close enough to swap out a layout engine?”. And I don’t think we have a good answer for that yet. We can try to do our best to find the set of non-conformant Yoga behavior to replicate, and then emulate that. But there’s such a wide variety of RN usage which could lose consistency in “quiet” ways that I would expect a long tail with a lot of breakage even if we were methodical. Runtime output diffing (throw if layouts disagrees) is one solution we can use during tests of real internal apps, but it’s too expensive to run in production applications. I also have some dream where we can take many real-world snapshots of nodes in an app (e.g. YGPrint output) so that we can catch changes to output and performance of some more real world cases in greater isolation. 

RN using Yoga also helps us to justify work to Yoga that affects users beyond RN. I don’t know what the long-term of that would look like if RN used an engine outside of Yoga.

Reply:
On ideology / project goals:

I'd say Taffy's goal is somewhat similar to Yoga's in that it intends to be a general-purpose layout library which is easily embeddable. Making this available to the Rust ecosystem is definitely a priority for us, but binding to other languages are definitely on our radar, and something we'd like to add. Where there is perhaps some divergence is that Taffy is keen on becoming layout-algorithm agnostic by defining a trait/interface that allows anybody to plug in their own layout algorithm (i.e. measure funcs on steroids). This is driven in large part by the wider Rust GUI community, not all of whom are keen on web-style layout. Whether this ends up being in conflict in with being fully web-compatible remains to be seen. But trying to allow interoperation with arbitrary layout modes while keeping the shared communication protocol keen certainly makes things more complicated.

On a personal note: my goal is really to enable a truly cross platform GUI solution (across all 5 major platforms: ios/android/windows/mac/linux). I feel like React Native is currently one of the best options for that (although it's state on windows and especially mac is frustrating). And Rust GUIs like [Druid/Xilem](https://github.com/linebender), [Iced](https://github.com/iced-rs/iced), and [Dioxus](https://dioxuslabs.com/)  also look promising in this space.

Reply:
> RN using Yoga also helps us to justify work to Yoga that affects users beyond RN. I don’t know what the long-term of that would look like if RN used an engine outside of Yoga.

Presumably the idea would be for that work to go into that other engine instead (which would also be open source and available for others to use). Although I suspect there are a significant number of people/projects already using Yoga who would like to maintain it anyway either because they don't want to switch APIs, or because they prefer a library written in C++. Is Meta unwilling to use and/or commit resources to projects that it does not solely maintain? If is it willing to do this, then I feel like the best thing for Yoga (regardless of whether RN continues to use it) would be to open up the maintenance to the wider community. There are clearly a lot of people willing to contribute (lot's of seemingly reasonable quality PRs), and it seems like the bottleneck is maintainer time. Are there any precedents for a more shared ownership model like this? If Meta isn't willing to go for a shared model, then I guess it would be best to make a clear decision: either commit to maintaining the library or release it to the community to maintain themselves.

Reply:
> Apart from the ideological goal, there’s a practical question of “how do we build enough confidence the behavior is close enough to swap out a layout engine?”.

Hmm... yes, I can definitely see that validating that a new implementation is truly compatible is a much harder task than simply adding a compatibility mode. I have to admit I am somewhat bemused by the focus on backwards compatibility. RN makes all sorts of breaking changes pretty much every release, although I suppose layout probably affects more code than most parts of RN (and also has the unfortunate property that it would likely silently render something different rather than failing).

I wonder if a transition to a new layout engine could be handled similar to the transition to "new architecture", with an opt-in flag at the application level. Library authors could then test and update their code to be compatible, and application developers the same. I guess the issue in this case is how much code would need to be updated at once? It could also be possible to ship both layout engines for a transition period, although I can definitely see that at that point it might make more sense just to stick to the original engine.

I would also bring up the point that if the goal is to be web compatible, then there will probably need to be breaking changes on an ongoing basis (albeit probably smaller than the initial diff), because web browsers themselves make breaking changes to align themselves with the spec, and any RN implementation will almost certainly contain spec-incompatible bugs. It's a fundamentally different approach to making the implementation definitive and keeping things 100% backwards compatible.

Reply:
Re compatibility, a big difference with other breaking changes are in signal and actionability. Most APIs we change can be detected statistically via type checking or the compiler, or scream loudly at runtime. So these are detectable. The nature of layout behavior changes is also tricky because it scales to every view of every app, instead of more confined scenarios. 

For companies with large existing codebases using RN, I don’t think we could tell folks “Try turning on this flag, and you might see UI changes, but we don’t know where”. Folks just wouldn’t switch, even if we could demonstrate compelling improvements. We would then need to support both layout engines in perpetuity. Though part of the “traceable styles” goal is that we can start shifting product code in the ecosystem still being developed which is reliant on broken behaviors. 

Fabric rollout has not yet been a success story in OSS. There are a lot of folks working now to instead start bridging the APIs, because users with large codebases have been reluctant to migrate. But there is a lot of effort to make that path easier. It reinforces that adding a flag to switch isn’t enough, and even after migrating a giant codebase internally, we are still finding bugs. 

Re browser breaking changes, it’s something I think we might be able to follow. There are some correctness changes like UseLegacyStretchBehavior that we know broke too many apps to enable for everyone, but Chrome changing behavior for the entire internet means some level of confidence they are not creating breaks everywhere. Really we need a way to know “how breaking” the change is before making it. One of the more powerful ways we have to do this is rolling the change out to a portion of Meta app users and monitoring if behavior changes. I.e. given a large enough set of code, does something break enough to shift user behavior. In those cases we can likely go ahead and make the break, though that measure is still lossy.

Reply:
Re ownership model, while Meta is still relying on Yoga as critical infrastructure, I think there will be a desire to retain ownership of it. React Native is definitely the main user of Yoga inside of Meta. But really getting off of Yoga, both for the thousands of surfaces inside Meta, and more in the community we can’t reach into and fix, ends up being very expensive, and prone to risk in fragmenting the paths we need to maintain without being able to remove the old path. These sort of large migrations really end up having an opportunity cost on what we could instead provide on the existing infrastructure, along with a cost to our users, where we’re really trying to focus now on making our upgrade story less painful. 

In a world where React Native wasn’t using Yoga, I don’t think it’s likely we would have someone to step in to maintain it. If we really didn’t have anyone inside of Meta using it or able to maintain it, I do like the idea of transferring ownership somewhere more shared, to allow the community to take a more active role in maintaining it. But I do genuinely believe RN using Yoga, and really wanting to invest in it, has the potential to help a project that is still being used throughout the industry.

Reply:
I still have worried a bit we could veer into fragmentation with StrictLayout, even if we put shiny new bits like grid behind it. It lets us make those breaking changes more confidently, and lets users fix code on a more granular level. E.g. even for a flex basis caching bug we wanted to fix, it was too expensive to do it app-wide in a relatively large app. But, my hope is that starting with a single layout engine which can alter its conformance level puts us in a position to sort of start priming the ecosystem for different behavior without yet taking the path of supporting multiple engines.